cpp/arcticdb/storage/failure_simulation.hpp

```diff
@@ -10,53 +10,143 @@
 #include <arcticdb/entity/protobufs.hpp>
 #include <arcticdb/util/random.h>
 #include <arcticdb/util/constructors.hpp>
+#include <arcticdb/util/preconditions.hpp>
+#include <arcticdb/util/variant.hpp>
+#include <folly/Function.h>
 
 namespace arcticdb {
 
+#ifdef _WIN32
+#undef DELETE
+#endif
+
 enum class FailureType : int {
     WRITE = 0,
-    READ
+    READ,
+    ITERATE,
+    DELETE,
 };
 
 static const char* failure_names[] = {
         "WRITE",
-        "READ"
+        "READ",
+        "ITERATE",
+        "DELETE",
 };
 
-struct FailureCategory {
-    double prob_;
-    bool thrown_;
+/** Function holder with a description. */
+struct FailureAction {
+    using FunctionWrapper = folly::Function<void(FailureType)>;
+    using Description = std::variant<const char*, std::string>;
+
+    Description description_;
+    FunctionWrapper::SharedProxy proxy_;
+
+    FailureAction(Description description, FunctionWrapper::SharedProxy proxy) :
+            description_(std::move(description)), proxy_(std::move(proxy)) {}
 
-    explicit FailureCategory(double prob) :  prob_(prob), thrown_(false) {}
+    template<typename Func>
+    FailureAction(Description description, Func func):
+        FailureAction(std::move(description), FunctionWrapper{func}.asSharedProxy()) {}
+
+    inline void operator()(FailureType type) const {
+        proxy_(type);
+    }
+};
+
+inline std::ostream& operator<<(std::ostream& out, const FailureAction& action) {
+    util::variant_match(action.description_,
+                        [&out](const char* desc) { out << desc ;},
+                        [&out](const std::string& desc) { out << desc ;});
```

Comment:
If this works it could be a bit simpler:
```suggestion
    std::visit([&out](const auto& desc){out << desc;}, action.description_);
```